Added A Test Policy (#4908)
status-im/status-go: _docs/policies/tests.md

# `status-go` Test Policy

- [Creating Tests](#creating-tests)
- [Flaky Tests](#flaky-tests)

## Creating Tests

- All new functionality MUST be introduced with tests that:
  - Prove that the functionality performs as described
  - Can be falsified
  - Are resistant to fuzzing

> - All new tests MUST BE validated via 1000 local tests. Ensuring that the test runs and passes consistently every time gives confidence that the test is not flaky.

Reviewer: This sentence I sort of disagree. Sort of because perhaps we should differentiate the type of test we want to rerun hundreds of times because a good portion of _unit tests_ are so deterministic in nature that this recommendation seems out of place or wasted effort to even worry about. Could we change the wording to something more flexible or realistic?

The tricky thing is that we haven't ever (or have we?) agreed on how to classify status-go's tests. If they all go in the same bucket as "tests" it becomes harder for us to express which ones we really want devs to increase `-count`.

Reviewer 2: I guess we should say that integration tests should be rerun

Author: That is reasonable. Let's change the language to:

```suggestion
- All new integration tests MUST BE validated via at least 10000 local tests. Ensuring that the test runs and passes consistently every time gives confidence that the test is not flaky.
```

I also agree with you @ilmotta we should have categories of test. Maybe we can open an issue and place a link to it in the Go Council page to mark this work as needing consensus.

Author: I'm also happy to add a definition of "integration tests."

@Samyoul make an issue and define

Author: This is done https://github.com/status-im/status-go/issues/4935

> - All new tests MUST BE validated via 1000 local tests. Ensuring that the test runs and passes consistently every time gives confidence that the test is not flaky.

Reviewer 3: it should be 10000 at least, 100000 is better though

Author: I'm happy to 10x the test count.

> - All new tests MUST BE validated via 1000 local tests. Ensuring that the test runs and passes consistently every time gives confidence that the test is not flaky.

Reviewer 4: Would be great to automate it and run on CI. What is not a CI check, will never be respected. Some will forget, some will not know (new contributors), some will not run to save time.

Author: I agree with your points. For now we can include it as part of the policy as this is the behaviour we want regardless of how we enforce the requirement.

I'll raise an issue to explore / implement the automation of "new integration test mega runs". I'll track it via the Go Council project board ... Which doesn't exist yet but we probably should have it as we have multiple improvements to track, and this improvement is dependent on being able to categorise certain tests as "integration test".

Author: @Samyoul make issue and somewhere to track Go Council related things

Author: Ok this issue is now done https://github.com/status-im/status-go/issues/4952

>     - add the `-count` flag to the test command eg: `-count 1000`

Reviewer 2: The name of the param changed to `test.count`
```suggestion
    - add the `-test.count` flag to the test command eg: `-test.count 1000`
```

Reviewer 4: How/where it changed? :thinking:
```
$ go help testflag
        -count n                                                                               
            Run each test, benchmark, and fuzz seed n times (default 1).
            If -cpu is set, run n times for each GOMAXPROCS value.                             
            Examples are always run once. -count does not apply to                             
            fuzz tests matched by -fuzz.
```

Author: I wasn't aware of this change, I checked the docs and I found one reference https://pkg.go.dev/testing#T.Parallel

> Parallel signals that this test is to be run in parallel with (and only with) other parallel tests. When a test is run multiple times due to use of -test.count or -test.cpu, multiple instances of a single test never run in parallel with each other.

But checking the latest version flag docs https://pkg.go.dev/cmd/go#hdr-Testing_flags I can see no reference to `-test.*`. 

It seems that `-test.*` just refers to the context of the flag rather than the actual usage of the flag 

---

EDIT: I'm not sure now. Seems that you can use both test flags `-count` and `-test.count`

    - OR wrap your test in a for loop
        ```go
        func TestTheThing(t *testing.T) {
            for i := 1; i < 1000; i++ {
                fmt.Println("Test Run", i)
                // your test body goes in here
            }

>         }

Reviewer 4: Please don't :sweat_smile:. We'll end up with some tests wrapped, others not. This should definitely be done either by using the `count` flag or a custom script.

Author: 😆 I should have added the 🙈 emoji and the made super clear that this is only for local usage. Nobody should try merging a test like that.

Reviewer 5: Actually I use this approach when I want a breakpoint to hit on a flaky test

Reviewer 6: > Actually I use this approach when I want a breakpoint to hit on a flaky test

never used this 🙂, i always use -test.count=1000 on flaky test and the breakpoint maybe get hit then

Author: I think the consensus here is to not recommend this approach. I'll remove this. Developers are free to dabble in the dark arts in the comfort of their own code editor.

        ```
## Flaky Tests

- All flaky tests / failing tests must be resolved

>   - Do not introduce flaky tests to the codebase

Reviewer 4: As mentioned earlier, ideally, this should be automated.

Author: I'll change the language to communicate that automation is the preference

> - Steps to resolving flaky test

Reviewer 5: Could we maybe do this a `h3.`? It's so many levels, hard to read. This way we could simplify it by one level 🙂

Reviewer 5: We could also add some decision diagram for better reading, but it's an overkill 😄

Reviewer 5: > We could also add some decision diagram for better reading, but it's an overkill 😄

well, @osmaczko was not as lazy as me I suppose 😄

Author: I'll do both

  - Ensure that the test is definitely flaky and that it wasn’t you that introduced a bug that made the test flaky.
      - Is a new test you’ve written flaky?

>         - Ok, you need to fix that before merge is acceptable.

Reviewer 4: Seems to be duplication of `Do not introduce flaky tests to the codebase` (I would remove the former).

Author: Yeah, this language is way too informal for a policy. I will change it.

      - Has an old test become flaky?

>         - Ok, did you touch this test or the functionality it tests? If yes, you need to fix that before merge is acceptable. Sorry.

Reviewer 4: If an old test becomes flaky, it will most likely make it through the CI unnoticed due to the retry mechanism. What we should do, is:

- Before merging, check rerun reports. If there are any, then, proceed as follows:
  - If the test does not appear in the flaky tests github issues or in the last three nightly test runs,
    - it is most likely that the flakiness was introduced by your changes and needs to be addressed before proceeding with the merge.
  - If the test is already documented as a flaky test (appears in the github issues or in the nightly test runs),
    - add a comment to the existing issue, or open a new issue if one does not already exist.

  - If an old test fails and seems flaky either locally or in CI, you must check the `status-go` GitHub repo issues for the test name(s) failing.
      - If the test appears in the list of flaky test issues
          - If the issue is open
              - Add a comment to the issue
              - Detail that you have experienced the test being flaky and in what context (local vs CI, link to the PR or branch).
          - If the issue is closed
              - Reopen the issue OR create a new issue referencing the previous issue
                - Either is fine, use your best judgement in this case.
              - Detail that you have experienced the test being flaky and in what context (local vs CI, link to the PR or branch).

>               - I just added this to see if you read the details of this PR `status-go` label for use `E:Flaky Test` may want to be not an epic, but I don't know

Reviewer 5: gotcha!
```suggestion
              - I just added `status-go` label for use `E:Flaky Test` may want to be not an epic, but I don't know
```

Reviewer 5: This is the only blocker of my review 😄

Reviewer 2: The whole line can be deleted, it makes no sense lol

Author: I'll delete it.

      - If the test does not appear in the list of flaky test issues
          - create a new issue
              - The issue title should include the flaky test name

>               - The issue should use the `E:Flaky Test` label, see here for all flaky tests https://github.com/status-im/status-go/labels/E%3AFlaky%20Test

Reviewer 5: I like how github automatically renders this label in markdown!
We could use it here directly instead of `E:Flaky Test`.

Author: I didn't realise that github did that, I really like it. I'll make this change and replace `E:Flaky Test` with https://github.com/status-im/status-go/labels/E%3AFlaky%20Test . Wow it is like magic.

# `status-go` Test Policy

- [Creating Tests](#creating-tests)
- [Flaky Tests](#flaky-tests)

## Creating Tests

- All new functionality MUST be introduced with tests that:
  - Prove that the functionality performs as described
  - Can be falsified
  - Are resistant to fuzzing

> - All new `integration tests` MUST BE validated via 10000 tests.

Reviewer 5: wow, I thought 10000 times was a joke.
I really think this is an overkill. 
If a test takes 5 seconds, then it would be 13 hours.

Author: Ok, fine. 1000

```suggestion
- All new `integration tests` MUST BE validated via a minimum of 1000 tests.
```

  - Ensuring that the test passes consistently every time gives confidence that the test is not flaky.

## Flaky Tests

Flaky tests are defined as tests that fail intermittently.

- All flaky tests / failing tests MUST be resolved.
- No flaky tests may be introduced into the codebase.

### Steps to resolving or reporting flaky tests

#### Is it me?
Determine who caused the flaky test.

- Is a new test you’ve written flaky or failing?
  - It was you.
  - You MUST fix the test before merge is acceptable.
- Has an existing test become flaky?
  - Check rerun reports. `TODO` add link to rerun reports
    - If the test does not appear in https://github.com/status-im/status-go/labels/E%3AFlaky%20Test or in the last three nightly test runs, it is most likely that the flakiness was introduced by your changes and needs to be addressed before proceeding with the merge.
    - Else the test is already documented as a flaky test (appears in the GitHub issues or in the nightly test runs), proceed to below.

```mermaid
flowchart TB
    A([PR ready for merge]) --> B{Have any test failed?}
    B -->|No| C[🎉 Proceed with merge 🪄]

>     B -->|Yes| D{Did you write the test for this PR?}

Reviewer 5: I suppose this should be "Did you write the test that is failing"?

Because if the answer to `Did you write the test for this PR?` is now, then you should always go and write one 😄

Author: > Did you write the test that is failing?

What if you wrote the test 2 years ago? We want to prevent new tests introduced by the PR the developer is trying to merge into `develop`.

Generally if you are introducing code, that code should come with tests. If you change existing code, tests that cover the existing code need to be considered for update.

What about this language?

```suggestion
    B -->|Yes| D{Is the failing test introduced or altered by this PR?}
```

Reviewer 5: yes this makes more sense to me, thank you 🙂

    D -->|No| E[Check rerun reports.]
    D -->|Yes| F[
        It is likely your changes introduced the flakiness.
        You MUST fix the test before merge is acceptable.
    ]
    F --> A
    E --> G{Does the test appear in `E:Flaky Test` issues<br/> or in the last three nightly test runs?<br/>}
    G -->|Yes| I[The flakiness needs reporting]
    G -->|No| F
    I --> J([Proceed to Reporting flow])
```

#### Reporting Flaky Tests

> - If an old test fails and/or seems flaky either locally or in CI, you MUST report the event.

Reviewer 5: All points below are sub-points to this one. Maybe just write it as text?

Author: Yeah, ok. That's reasonable, I'll collapse those nested lists one more level.
